hypothesis-python/src/hypothesis/internal/conjecture/shrinker.py

```diff
@@ -1330,9 +1326,63 @@ def lower(k):
 
         find_integer(lower)
 
+    def minimize_nodes(self, nodes):
+        ir_type = nodes[0].ir_type
+        value = nodes[0].value
+        # unlike ir_type and value, kwargs are *not* guaranteed to be equal among all
+        # passed nodes. We arbitrarily use the kwargs of the first node. I think
+        # this is unsound (= leads to us trying shrinks that could not have been
+        # generated), but those get discarded at test-time, and this enables useful
+        # slips where kwargs are not equal but are close enough that doing the
+        # same operation on both basically just works.
+        kwargs = nodes[0].kwargs
+        assert all(
+            node.ir_type == ir_type and ir_value_equal(ir_type, node.value, value)
+            for node in nodes
+        )
+
+        if ir_type == "integer":
+            # try shrinking from both sides towards shrink_towards
+            shrink_towards = kwargs["shrink_towards"]
+            Integer.shrink(
+                abs(shrink_towards - value),
+                lambda n: self.try_shrinking_nodes(nodes, shrink_towards + n),
+            )
+            Integer.shrink(
+                abs(shrink_towards - value),
+                lambda n: self.try_shrinking_nodes(nodes, shrink_towards - n),
+            )
```

Comment:
This is pretty brute force and does ~double the shrinks necessary in the worst case.

I think master behavior for integer shrinking is something like "shrink towards 0 on the current sign (only positive or only negative) and occasionally try and flip the sign bit". I think master gets a bit luckier/smarter than this in practice, because a straight-line translation of that here shrinks `minimal(st.integers(-5, 5), lambda n: not (-1 < n < 2))` to 2 instead of -1; to find -1 it needs to pursue the worse shrink of 2 to -2 and then shrink -2 to -1.

Reply:
> to find -1 it needs to pursue the worse shrink of 2 to -2 and then shrink -2 to -1

I suspect it's getting lucky with some sort of punning here, as the `integers` strategy is a mixture under the hood. I don't think it can reliably do this strategy - there's no backtracking in the shrinker - but it might be doing something that causes previous bytes for a larger integer width to be reinterpreted as a negative integer.

Reply:
yeah, good point. master fails on a trivial extension of [this problem](https://github.com/HypothesisWorks/hypothesis/pull/3962#discussion_r1586899822):

```python
assert minimal(st.integers(101, 200) | st.integers(0, 100), lambda n: n not in [101, 102, 0]) == 103
```

so it's just getting lucky.